hash_reader.go

```diff
@@ -22,22 +22,71 @@ import (
 	"github.com/amzn/ion-go/ion"
 )
 
-// HashReader is meant to share the same methods as the ion.Reader and hashValue interfaces.
-// However embedding both ion.Reader and hashValue results in a duplicate method build error because both interfaces
-// have IsNull(), Type(), and IsInStruct() methods defined.
-// So we embed the ion.Reader interface and explicitly list the remaining hashValue methods to avoid the error.
-// HashReader also provides a Sum function which allows read access to the hash value held by this reader.
+// A HashReader reads a stream of Ion values and calculates the hash.
+//
+// The HashReader has a logical position within the stream of values, influencing the
+// values returned from its methods. Initially, the HashReader is positioned before the
+// first value in the stream. A call to Next advances the HashReader to the first value
+// in the stream, with subsequent calls advancing to subsequent values. When a call to
+// Next moves the HashReader to the position after the final value in the stream, it returns
+// false, making it easy to loop through the values in a stream.
+//
+// 	var r HashReader
+// 	for r.Next() {
+// 		// ...
+// 	}
+//
+// Next also returns false in case of error. This can be distinguished from a legitimate
+// end-of-stream by calling Err after exiting the loop.
+//
+// When positioned on an Ion value, the type of the value can be retrieved by calling
+// Type. If it has an associated field name (inside a struct) or annotations, they can
+// be read by calling FieldName and Annotations respectively.
+//
+// For atomic values, an appropriate XxxValue method can be called to read the value.
+// For lists, sexps, and structs, you should instead call StepIn to move the HashReader in
+// to the contained sequence of values. The HashReader will initially be positioned before
+// the first value in the container. Calling Next without calling StepIn will skip over
+// the composite value and return the next value in the outer value stream.
+//
+// At any point while reading through a composite value, including when Next returns false
+// to indicate the end of the contained values, you may call StepOut to move back to the
+// outer sequence of values. The HashReader will be positioned at the end of the composite value,
+// such that a call to Next will move to the immediately-following value (if any).
+//
+// Sum will return the hash of the entire stream of Ion values that have been looped through.
+//
+//  cryptoHasherProvider := NewCryptoHasherProvider(SHA256)
+// 	r := NewTextReaderStr("[foo, bar] [baz]")
+//  hr := NewHashReader(r, cryptoHasherProvider)
+// 	for hr.Next() {
+// 		if err := hr.StepIn(); err != nil {
+// 			return err
+// 		}
+// 		for hr.Next() {
+// 			fmt.Println(hr.StringValue())
+// 		}
+// 		if err := hr.StepOut(); err != nil {
+// 			return err
+// 		}
+// 	}
+// 	if err := hr.Err(); err != nil {
+// 		return err
+// 	}
+//
+//  fmt.Printf("%v", hr.sum(nil))
```

Comment:
I think for code blocks we need a leading 4 spaces to each line and godoc will render them as code https://golang.org/src/encoding/gob/doc.go

Reply:
Updated hash_reader and hash_writer